---
title: "Chart Combinations"
author: "Anamaria Crisan"
date: '2019-01-09'
output: html_document
---


This document demonstrates how to combine and plot different chart types using gevitR. This document assumes that you have already run the "Getting started with gevitR" and the "Simpel Chart Types" and that the necessary data as already been loaded into your R workspace.

Another good option for data: https://datadryad.org/bitstream/handle/10255/dryad.111738/ZIKV_with1966_strict_skyline_ctmc.xml?sequence=1
Also exists in microreact

## Loading data
```{r,message=FALSE,warning=FALSE}
devtools::load_all() #temporary once things are done
library(dplyr)

#Table data
tab_dat<-input_data(file = system.file("extdata", "ebov_metadata.csv", package = "gevitR"),dataType = "table")

#Tree data
tree_dat<-input_data(file = system.file("extdata", "ebov_tree.nwk", package = "gevitR"),dataType = "tree")

#Genomic data
genomic_dat<-input_data(file = system.file("extdata", "ebov_GIN_genomic.fasta", package = "gevitR"),dataType = "dna")

#Shape files
#Shape files require that .shp,.shx,and .prj files at a minimun to be in the same directory
#to add metadata to the shape file, you can also add .dbf files
gin_shape_dat<-input_data(file = system.file("extdata", "gin_admbnda_adm1_ocha_itos.shp", package = "gevitR"),dataType = "spatial")
lbr_shape_dat<-input_data(file = system.file("extdata", "lbr_admbnda_adm1_ocha.shp", package = "gevitR"),dataType = "spatial")
sle_shape_dat<-input_data(file = system.file("extdata", "sle_admbnda_adm1_1m_gov_ocha_20161017.shp", package = "gevitR"),dataType = "spatial")
```

## Chart Combinations

First, let's create a few charts that can be used
```{r individual_charts,message=FALSE,warning=FALSE}
bar_chart<-specify_base(chart_type = "bar",data="tab_dat",x = "country")

phyloTree_chart<-specify_base(chart_type = "phylogenetic tree",data="tree_dat")

scatter_chart<-specify_base(chart_type = "scatter",data="tab_dat",x = "latitude",y="longitude",title = "Ebola Scatter Plot")

```

### Many Types General
Many types general are essentially when the you just want to put a bunch of plots together and there are no spatial or visual linkages between the plots themselves.

```{r many_types_general,message=FALSE,warning=FALSE}
mg_combo<-specify_combination(combo_type = "many_types_general",base_charts = c("bar_chart","phyloTree_chart","scatter_chart"))
plot(mg_combo)
```

### Small Multiples
Small multiple charts are visually linked because the show the same underlying chart type, however they show different aspects of the data.

```{r small_multiples,message=FALSE,warning=FALSE}
#set up the base chart type with all the data that you wish to use
scatter_chart<-specify_base(chart_type = "scatter",data="tab_dat",x = "latitude",y="longitude")

#now specify the small multiple combination
sm_combo_scatter<-specify_combination(combo_type = "small_multiple",base_charts = "scatter_chart",facet_by = "country")
plot(sm_combo_scatter)

sm_combo_bar<-specify_combination(combo_type = "small_multiple",base_charts = "bar_chart",facet_by = "country")
plot(sm_combo_bar)

#let's try a more interesting bar_chart small multiple
bar_chart_alt<-specify_base(chart_type = "bar",data="tab_dat",x = "year",title="All together")
sm_combo_bar_alt<-specify_combination(combo_type = "small_multiple",base_charts = "bar_chart_alt",facet_by = "country")
plot(bar_chart_alt)
plot(sm_combo_bar_alt)

```

Other chart types cannot be easily subsetted. For example, it's not straight forward to make a small multiple of a phylogenetic tree while only showing a subset of the tree, the same is generally true for maps, and networks. Yes there are ways to truely subset them, but it's messy and in the whole network structure matters, so here you get the whole network structure.

In the current implementation of gevitR, there needs to be some tablular data assocaited with non-tabular data in order to be able to understand what should be visualized in the first place.

```{r small_multiples_complex,message=FALSE,warning=FALSE}
#Tree data
tree_dat<-input_data(file = system.file("extdata", "ebov_tree.nwk", package = "gevitR"),dataType = "tree")

tree_dat_meta<-input_data(file = system.file("extdata", "ebov_tree.nwk", package = "gevitR"),dataType = "tree",metadata = system.file("extdata", "ebov_metadata.csv", package = "gevitR"))

tree_chart<-specify_base(chart_type = "phylogenetic tree",data="tree_dat_meta")
plot(tree_chart)

sm_combo_tree<-specify_combination(combo_type = "small_multiple",base_charts = "tree_chart",facet_by = "country")
plot(sm_combo_tree)

```

### Composite combinations

*This example will work without any errors*
```{r composite_no_error,message=FALSE,warning=FALSE}
phyloTree_chart<-specify_base(chart_type = "phylogenetic tree",data="tree_dat")
scatter_chart<-specify_base(chart_type = "scatter",data="tab_dat",x = "month",y="site.id")
scatter_chart_two<-specify_base(chart_type = "scatter",data="tab_dat",x = "country",y="site.id")


composite_combo<-specify_combination(combo_type = "composite",
                                     base_charts = c("phyloTree_chart","scatter_chart","scatter_chart_two"))

plot(composite_combo)

```

Example without the phylogenetic tree


*This example will have an error because pie charts cannot be part of composite combinations*
Pie chart should be autonatically removed from the specifications
```{r composite_no_error,message=FALSE,warning=FALSE,}

pie_chart<-specify_base(chart_type = "pie",data="tab_dat",x = "country")
composite_combo<-specify_combination(combo_type = "composite",
                                     base_charts = c("phyloTree_chart","scatter_chart","scatter_chart_two","pie_chart"))


plot(composite_combo)

```

```{r}
bar_alt<-specify_base(chart_type = "bar",data="tab_dat",x = "site.id",y="month",rm_x_label=TRUE)

composite_combo<-specify_combination(combo_type = "composite",
                                     base_charts = c("phyloTree_chart","bar_alt","scatter_chart_two"))

plot(composite_combo)

```

*It also works when there isn't a tree invovled.
```{r}
composite_combo<-specify_combination(combo_type = "composite",
                                     base_charts = c("scatter_chart","bar_alt","scatter_chart_two"))

plot(composite_combo)

```
### Color Linked Combinations

```{r}
ebov<-tab_dat@data[[1]] #make this nicer
ebov_case_counts<-ebov %>% 
  group_by(country,month) %>%
  count()

line_chart<-specify_base(chart_type = "line",data="ebov_case_counts",x = "month",y="n",group = "country")

#let's link these results to a scatter chart and the phylogenetic tree

```
